### PR TITLE
Add support for setting subject configuration

### PR DESCRIPTION
**Pull request by bspeakmon**
This patch adds support for configuring subjects in the registry, which currently involves setting the compatibility level to be enforced on subsequent schema registrations to a subject.

I needed this for an internal project.

### CODE REVIEW

File: src/main/kotlin/com/github/imflog/schema/registry/config/ConfigSubjectExtension.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.github.imflog.schema.registry.config
+
+open class ConfigSubjectExtension {
+    val subjects: ArrayList<Pair<String, String>> = ArrayList()
+
+    fun subject(inputSubject: String, compatibility: String) {
```
Comment (ImFlog) [inline src/main/kotlin/com/github/imflog/schema/registry/config/ConfigSubjectExtension.kt]: I think we should use the https://github.com/confluentinc/schema-registry/blob/master/client/src/main/java/io/confluent/kafka/schemaregistry/avro/AvroCompatibilityLevel.java
Like this we can use static typing in the object definitions even if in the end we have to use the `.name` of the enum in our rest call.

File: src/test/kotlin/com/github/imflog/schema/registry/config/ConfigTaskActionTest.kt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        )
+
+        // when
+        val errorCount = ConfigTaskAction(
+            registryClient,
+            subjects
+        ).run()
+
+        // then
+        Assertions.assertThat(errorCount).isEqualTo(0)
```
Comment (ImFlog) [inline src/test/kotlin/com/github/imflog/schema/registry/config/ConfigTaskActionTest.kt]: If we use the AvroCompatilibityLevels, we should test that we call the mockClient with the right parameters (using mockk verify).

**Comment by ImFlog** [comment]
I am wondering if we shouldn't add this compatibility setting in the register task too.
We would call it after the register to allow user to change it directly if specified.
WDYT ?

File: src/main/kotlin/com/github/imflog/schema/registry/config/ConfigSubjectExtension.kt (hunk 1 above)
Comment (bspeakmon) [reply src/main/kotlin/com/github/imflog/schema/registry/config/ConfigSubjectExtension.kt]: Annoyingly, this isn't possible. The class has a public `name` field and a public method `name`, and Kotlin can't distinguish between them. What I will do is `valueOf()` the `String` to make sure it's an acceptable compatibility level before I pass it into the task.

File: src/test/kotlin/com/github/imflog/schema/registry/config/ConfigTaskActionTest.kt (hunk 2 above)
Comment (bspeakmon) [reply src/test/kotlin/com/github/imflog/schema/registry/config/ConfigTaskActionTest.kt]: Good idea, but as above, not possible.

**Comment by bspeakmon** [comment]
I had the same thought and decided it wasn't the right thing to do.

Part of the idea of registering schemas is that the registration _should_ fail if the compatibility is invalid. If the same entity is both consumer and producer of a schema, that's fine; compatibility breaks only affect them. But if the compatibility is a contract between the schema producer and consumer as two different entities, then the producer should _not_ be able to change that at will, lest they break consumers without warning.

In our case, we have people iterating on a producer schema at will and many consumers of it. As long as the compatibility level is consistent, both sides are protected: change can happen at will. Registering a new .avsc as a subject does not give one the right to change the compatibility contract, however.

**Comment by ImFlog** [comment]
I thought about it too and you are right, this should not be in the register task.
What I have in mind is something like a `createSubjectTask` that create a subject and set the required compatibility. A task that would fail if the subject already exist.

I am not sure it impacts this code but I have question about what is the use case of such task ?

**Comment by bspeakmon** [comment]
I could think of one possible use case.

* Local development/test cases or DR process: existing code repository with schemas, brand new deployment with a schema registry. Run the build once to set everything up. Instead of writing a curl script in bash, you have a platform-independent tool for doing this. Existing subjects would be an error, so you can't accidentally clobber something important.

**Comment by ImFlog** [comment]
Don't you think It would be better to replace this task by one that create a subject and then set compatibility that could only run once ? To avoid compatibility switch all the time ?

**Comment by bspeakmon** [comment]
I think not. At least in my use case, it's possible that I want to change compatibility later on. I may want to start at `FULL_TRANSITIVE` and later relax that to `BACKWARD_TRANSITIVE`, which is fully compatible with all existing subject versions. With a separate config task, I can do that while documenting my configuration intention separately from my schema registration action. 

My use case isn't the only one though. I can put together a task like you describe if you feel that the other use case (of an immutable, first-time setup) is important for you :)

**Comment by ImFlog** [comment]
No worries, I'll merge this and work on the one I talked about a bit later :)
I'll release this tommorow I think.
Thank you again !

**Comment by bspeakmon** [comment]
Sounds good! Thanks so much for taking the patch!

**Comment by ImFlog** [comment]
The version 0.7.0 is now available in the gradle portal ;)

**Pull request merged**